### PR TITLE
cpu: aarch64: replace acl with kleidiai

### PR DESCRIPTION
**Pull request by jondea**
See related RFC: https://github.com/uxlfoundation/oneDNN/pull/5145

### CODE REVIEW

File: src/cpu/aarch64/kai_convolution.cpp [hunk 1]
```diff
@@ -0,0 +1,746 @@
+/*******************************************************************************
+* Copyright 2020-2026 Arm Ltd. and affiliates
```
Comment (renato-arantes) [inline src/cpu/aarch64/kai_convolution.cpp]: Why 2020-2026? Shouldn't it be only 2026?

File: src/cpu/aarch64/kai_utils.hpp [hunk 2]
```diff
@@ -0,0 +1,190 @@
+/*******************************************************************************
+* Copyright 2023-2026 Arm Ltd. and affiliates
```
Comment (renato-arantes) [inline src/cpu/aarch64/kai_utils.hpp:2]: Same here: Shouldn't be only 2026?

File: src/cpu/aarch64/matmul_inner_product.hpp [hunk 3]
```diff
@@ -0,0 +1,81 @@
+/*******************************************************************************
+* Copyright 2025 Intel Corporation
```
Comment (renato-arantes) [inline src/cpu/aarch64/matmul_inner_product.hpp:2]: Why do we have to include Intel's copyright in those new files?

File: src/cpu/cpu_deconvolution_list.cpp [hunk 4]
```diff
@@ -69,7 +66,6 @@ const std::map<pk_impl_key_t, std::vector<impl_list_item_t>> &impl_list_map() {
             CPU_INSTANCE_AARCH64(brgemm_deconvolution_fwd_t<sve_256>)
             CPU_INSTANCE_AARCH64(brgemm_deconvolution_fwd_t<sve_128>)
             CPU_INSTANCE_AARCH64(jit_sve_512_core_x8s8s32x_deconvolution_fwd_t)
-            CPU_INSTANCE_AARCH64_ACL(acl_deconvolution_fwd_t)
```
Comment (renato-arantes) [inline src/cpu/cpu_deconvolution_list.cpp:72]: We need an ASIMD implementation.

File: src/cpu/cpu_prelu_list.cpp [hunk 5]
```diff
@@ -41,7 +38,6 @@ const std::map<pk_impl_key_t, std::vector<impl_list_item_t>> &impl_list_map() {
     static const std::map<pk_impl_key_t, std::vector<impl_list_item_t>> the_map = REG_PRELU_P({
         {{forward}, {
             CPU_INSTANCE_X64(jit_prelu_fwd_t)
-            CPU_INSTANCE_AARCH64_ACL(acl_prelu_fwd_t)
```
Comment (renato-arantes) [inline src/cpu/cpu_prelu_list.cpp]: Also here: we need an AArch64 implementation.

File: src/cpu/aarch64/kai_convolution.cpp (hunk 1 above)
Comment (jondea) [reply src/cpu/aarch64/kai_convolution.cpp]: Significant content was taken from acl convolution, hence the backdating

File: src/cpu/aarch64/kai_utils.hpp (hunk 2 above)
Comment (jondea) [reply src/cpu/aarch64/kai_utils.hpp:2]: Significant content was taken from acl_utils.*, hence the backdating

File: src/cpu/aarch64/matmul_inner_product.hpp (hunk 3 above)
Comment (jondea) [reply src/cpu/aarch64/matmul_inner_product.hpp:2]: Significant content was taken from src/cpu/x64/matmul_inner_product.hpp, hence the copyright.

File: src/cpu/cpu_deconvolution_list.cpp (hunk 4 above)
Comment (jondea) [reply src/cpu/cpu_deconvolution_list.cpp:72]: Agreed!

**Comment by cfRod** [comment]
Does the coverage increase / decrease compared to before?

**Comment by jondea** [comment]
This change should not cause anything to go from an optimized to an unoptimized implementation. If you spot any (particularly ones at the framework level!) then it is a bug, and needs reporting here. Known issues are:
- prelu
- asimd log/log softmax
- MMLA s8:s8:s8
- gemm/conv + any post ops apart from relu (alpha=0) or sum
- batch norm with channels not a multiple of 16 
I think I discuss most of these in [the RFC](https://github.com/uxlfoundation/oneDNN/pull/5145), and we intend to get these fixed before merging. No doubt I have missed some though, so bug reports are very welcome here!

There will also be some things that go from an unoptimized to an optimized implementation, but they are mostly
- edge cases which become easier to handle with `kai` ops because a layer of API has been removed
- most of the JIT impls are more flexible than their ACL counterparts, so expanding their use tends to unintentionally bring extra coverage

It's worth saying that this change gives us more flexibility and will let us increase coverage more easily afterwards, however, it isn't the primary purpose of this patch itself. But, if you have specific and straightforward asks e.g. s8s8s32 we may be able to easily enable them in this patch.

File: src/cpu/aarch64/kai_convolution.cpp [hunk 6]
```diff
@@ -0,0 +1,746 @@
...
+}
+
+bool can_use_direct_1x1_src(const kai_convolution_fwd_t::pd_t &pd) {
+    return !pd.indirect_ && exact_src_dst_tag(*pd.src_md()) == format_tag::nhwc
+            && pd.KH() == 1 && pd.KW() == 1 && pd.KSH() == 1 && pd.KSW() == 1
+            && pd.padT() == 0 && pd.padL() == 0 && pd.OH() == pd.IH()
+            && pd.OW() == pd.IW();
+}
+
+int cap_threads_for_small_gemm(
```
Comment (fadara01) [inline src/cpu/aarch64/kai_convolution.cpp]: We should be cautious with capping threads. In micro-benchmarks or benchdnn, capping threads will likely show nice speedups. However, this can be problematic when you benchmark models end to end, due to scenarios like: 
- Max threads is 64
- We have a model with Layer A and Layer B
- Layer A has threads capped from 64 to 32
- As a result, OMP runtime might shrink the team size (from 64 to 32) - I say might because it's up to the OMP implementation: libomp doesn't tend to do this, but libgomp does this more aggressively.
- Layer B needs all 64 threads, as a result we have to pay the overhead of re-spawningg the 32 threads destroyed in the previous step.

See https://github.com/vllm-project/vllm/issues/34229 and https://github.com/uxlfoundation/oneDNN/pull/4679 for practical examples where this bit us before.

I assume we're now relying on `dnnl::prarallel_for` in place of ACL's `OMPScheduler`. When doing something like `parallel(std::min(pd()->mm_parallel_work, num_threads)`, with libgomp, the team size will shrink from `num_threads` to `mm_parallel_work` as this call becomes: `#pragma omp parallel num_threads(mm_parallel_work)`.

The same pattern wouldn't cause team size to shrink with ACL's `OMPScheduler` as it keeps the extra threads (`omp_num_threads - amount_of_work`) alive and just schedules 0 iterations on them:
```
#pragma omp parallel for firstprivate(info) num_threads(omp_num_threads) default(shared) proc_bind(close) \
    schedule(static, 1)
    for (unsigned int wid = 0; wid < amount_of_work; ++wid)
    {
        info.thread_id = wid;
        workloads[wid](info);
    }
}
```

File: src/cpu/aarch64/matmul/kai_matmul.cpp [hunk 7]
```diff
@@ -0,0 +1,509 @@
...
+    if (get_verbose(verbose_t::profile_externals)) {
+        std::cout << "profile_externals: " << _kernel->get_config().filter
+                  << std::endl;
+    }
+
+    const auto scratchpad = ctx.get_scratchpad_grantor();
+
+    const kai::ops::ndrange_t window_size = _kernel->get_window_size();
+    const int num_windows = static_cast<int>(window_size.total_size());
+    int num_threads = std::min(num_windows, dnnl_get_current_num_threads());
```
Comment (fadara01) [inline src/cpu/aarch64/matmul/kai_matmul.cpp]: See other comment about this.

**Review by fadara01 — COMMENTED** [review]
This is awesome!

I didn't have a very close look, just commented on some concerns about changes in threading behavior

File: src/cpu/aarch64/kai_convolution.cpp (hunk 6 above)
Comment (jondea) [reply src/cpu/aarch64/kai_convolution.cpp]: Excellent point, I have fixed this now.

**Comment by cfRod** [comment]
Can you exclude all the kai implementations from threadpool using " VDISPATCH_MATMUL(DNNL_CPU_THREADING_RUNTIME != DNNL_RUNTIME_THREADPOOL,
            VERBOSE_UNSUPPORTED_RUNTIME);" until we can make them async compatible
            
 See this PR https://github.com/uxlfoundation/oneDNN/pull/4297

File: src/cpu/aarch64/matmul/kai_matmul.cpp [hunk 8]
```diff
@@ -0,0 +1,605 @@
...
+            kernel_execute_work(*pd()), max_threads);
+
+    std::unique_ptr<kai::ops::IGemmCommon> _kernel = nullptr;
+    if (pd()->is_dequant()) {
+        DEFINE_ARG_SCALES_BUFFER(src_scale, DNNL_ARG_SRC);
+        DEFINE_ARG_SCALES_BUFFER(wei_scale, DNNL_ARG_WEIGHTS);
+        kai::ops::DequantizeFloat dequant(src_scale[0] * wei_scale[0]);
+        _kernel = pd()->create_kai_gemm_dequant(dequant, num_threads);
+    } else {
+        _kernel = pd()->create_kai_gemm(num_threads);
```
Comment (renato-arantes) [inline src/cpu/aarch64/matmul/kai_matmul.cpp:459]: I understand why you need to recreate the kernel inside `execute()` when quantisation is needed, but I think the `else` is redundant. If `num_threads` is needed, you can also get it inside `init` [here](https://github.com/uxlfoundation/oneDNN/blob/c03026f088f8d53ad53ed375930efcbf55dffb42/src/cpu/aarch64/matmul/kai_matmul.cpp#L323).
